lib/twocohom.gi

```diff
@@ -1590,20 +1697,221 @@ local r,z,ogens,n,gens,str,dim,i,j,f,rels,new,quot,g,p,lay,m,e,fp,old,sim,
       hom:=GroupHomomorphismByImages(r.group,Group(r.module.generators),
         GeneratorsOfGroup(r.group),r.module.generators);
       p:=Image(quot);
+      trysy:=Maximum(1000,50*IndexNC(p,SylowSubgroup(p,prime)));
+      # allow to enforce test coverage
+      if ValueOption("forcetest")=true then trysy:=2;fi;
+
+      mindeg:=2;
+      if IsPermGroup(p) then
+        mindeg:=Minimum(List(Orbits(p,MovedPoints(p)),Length));
+      fi;
       while Size(p)<Size(fp) do
-        it:=DescSubgroupIterator(p);
+        # we allow do go immediately to normal subgroup of index up to 4.
+        # This reduces search space
+        it:=DescSubgroupIterator(p:skip:=LogInt(Size(p),2));
         repeat
           m:=NextIterator(it);
           e:=fail;
-          if hom=false or Size(m)=1 or
+          if Index(p,m)>=mindeg and (hom=false or Size(m)=1 or
             false<>MatricesStabilizerOneDim(r.module.field,
               List(GeneratorsOfGroup(m),
-              x->TransposedMat(ImagesRepresentative(hom,x))^-1)) then
+              x->TransposedMat(ImagesRepresentative(hom,x))^-1))) then
             Info(InfoExtReps,2,"Attempt index ",Index(p,m));
-            e:=LargerQuotientBySubgroupAbelianization(quot,m);
+            if hom=false then Info(InfoExtReps,2,"hom is false");fi;
+
+            if hom<>false and Index(p,m)>
+              # up to index
+              50
+              # the rewriting seems to be sufficiently spiffy that we don't
+              # need to worry about this more involved process.
+              then
+
+              # Rewriting produces a bad presentation. Rather rebuild a new
+              # fp group using rewriting rules, finds its abelianization,
+              # and then lift that quotient
+              sub:=PreImage(quot,m);
+              tab0:=AugmentedCosetTableInWholeGroup(sub);
```

Comment:
Observation: `this new code starting at line 1732 ` is never executed by the Travis test suite